[Fix] `install`: detect user shell and try shellrc file first

steelcowboy opened the pull request:
# Situation
Currently, the nvm install script has a specific order in which it looks for files:
1. `.bashrc`
2. `.bash_profile`
3. `.zshrc`

However, this doesn't take into account any kind of user preference. I suggest we first try and see if we can find the user's shell and determine a good RC file to use based on that. If not, we fall back to the current behavior.

# Tests
```
> nvm@0.35.3 test/install_script /home/steelcowboy/github/nvm
> shell=$(basename -- $(ps -o comm= $(ps -o ppid= -p $PPID)) | sed 's/^-//'); make TEST_SUITE=install_script test-$shell


Running tests in zsh
Running tests at 2020-08-08T18:49:44
  test/install_script
  ✓ nvm_check_global_modules
  ✓ nvm_detect_profile
  ✓ nvm_do_install
  ✓ nvm_install_dir
  ✓ nvm_install_with_aliased_dot
  ✓ nvm_install_with_node_version
  ✓ nvm_profile_is_bash_or_zsh
  ✓ nvm_reset
  ✓ nvm_source

Done, took 11 seconds.
9 tests passed.
0 tests failed.
```
I had to change the shell on `nvm_install_with_aliased_dot` to bash in order to use `shopt`, open to a different approach though!

# Current Behavior
```
steelcowboy@asclepius:~/bin|⇒  echo $SHELL
/bin/zsh
steelcowboy@asclepius:~/bin|⇒  bash -c 'echo $SHELL'
/bin/zsh
steelcowboy@asclepius:~/bin|⇒  ./nvm-install.sh 
=> Downloading nvm from git to '/home/steelcowboy/.nvm'
=> Cloning into '/home/steelcowboy/.nvm'...
...
Resolving deltas: 100% (35/35), done.
=> Compressing and cleaning up git repository

=> Appending nvm source string to /home/steelcowboy/.bashrc
=> Appending bash_completion source string to /home/steelcowboy/.bashrc
=> Close and reopen your terminal to start using nvm or run the following to use it now:

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
```

# New Behavior
```
steelcowboy@asclepius:~/github/nvm|install-script-fix-zsh ⇒  curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13527  100 13527    0     0  53678      0 --:--:-- --:--:-- --:--:-- 53678
=> Downloading nvm from git to '/home/steelcowboy/.nvm'
=> Cloning into '/home/steelcowboy/.nvm'...
...
Resolving deltas: 100% (35/35), done.
=> Compressing and cleaning up git repository

=> Appending nvm source string to /home/steelcowboy/.zshrc
=> Appending bash_completion source string to /home/steelcowboy/.zshrc
=> Close and reopen your terminal to start using nvm or run the following to use it now:

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
```
# Docker Testing
## bash
https://gist.github.com/steelcowboy/451d45ae6a1728dba7678ce2e68dbcc4
## zsh
https://gist.github.com/steelcowboy/1d7ac8aed8e7b6a275aa1d35ac43cd1a
## dash
https://gist.github.com/steelcowboy/1af21c23cac3208b9dc77676475e65d3
## ksh
https://gist.github.com/steelcowboy/5b62b7ac44c7e425c504a86ba511d1dc
*Note*: It doesn't like the `local` in nvm.sh, but this is a known issue: #574 
## mksh
https://gist.github.com/steelcowboy/5b62b7ac44c7e425c504a86ba511d1dc
Interestingly unlike AT&T ksh, mksh is perfectly happy with the local variables :) 
# MacOS Issue
Resolves #592, but #2148 still remains because Catalina doesn't create any default .zshrc it would seem. If we wanted to take the spirit of this change further, we could simply `touch $SHELLRC` because if your shell is set I'd imagine there's a high likelihood you want an rc file for it :)

I think if we do want to go that direction, the best approach is to do it in another PR to keep the changes well-scoped.

ljharb commented on install.sh:
```diff
@@ -233,7 +233,18 @@ nvm_detect_profile() {
+  local SHELLRC
+
+  DETECTED_SHELL="${SHELL##*/}"
+  SHELLRC="$HOME/.${DETECTED_SHELL}rc"
```
how does this affect bashrc vs bash_profile, zshrc vs zprofile, etc?

steelcowboy replied:
So with this, the current logic reads:
1. Get the user's shell
2. If they have .bashrc or .zshrc, use that
3. Otherwise, check if BASH_VERSION is set
4. If yes, check for .bashrc, then .bash_profile

The question is: do we want mirror logic with DETECTED_SHELL. I'm taking advantage of a single logical branch: trying SHELLRC, which is likely most common, then iterating through various other files. However, [based on this note about the $SHELL variable](http://tldp.org/LDP/abs/html/internalvariables.html#AMIROOT), it's not builtin and I'm not sure if it's a system guarantee.

I'm not really sure how thorough we want to be here, what do you think?

ljharb replied:
I definitely think we can't rely on $SHELL being present. I'm also not sure if `.dashrc` or `.kshrc` are even things.

This logic should probably be as verbose as needed to be very clear about what it's doing and in what order. I'd like to be as thorough as possible, and in particular, I don't want to break anything that's currently working.

ljharb commented on install.sh:
```diff
@@ -233,7 +233,18 @@ nvm_detect_profile() {
+  DETECTED_SHELL="${SHELL##*/}"
+  SHELLRC="$HOME/.${DETECTED_SHELL}rc"
+
+  if [ -n "${DETECTED_SHELL}" ]; then
+    if [ -f "$SHELLRC" ]; then
+      DETECTED_PROFILE="$SHELLRC"
+    fi
```
```suggestion
  if [ -n "${DETECTED_SHELL}" ] && [ -f "$SHELLRC" ]; then
    DETECTED_PROFILE="$SHELLRC"
```
otherwise, when SHELLRC doesn't exist, it won't fall back to the bash/zsh detection cases.

steelcowboy replied:
Good catch!

ljharb commented on test/install_script/nvm_install_with_aliased_dot:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
```
why does this one need to change?

steelcowboy replied:
It's because of the `shopt` bash builtin being used. Here's the result of running that test suite without this:
```
  ✗ nvm_install_with_aliased_dot
setup:1: command not found: shopt
nvm
cleanup:3: command not found: shopt
```

steelcowboy commented:
Hey @ljharb, I accepted a few of your suggestions and added some context for the sh-bash change. What more do we need to move this forward?

ljharb commented:
@steelcowboy sorry for the delay; replied to your comment.

steelcowboy commented:
Hey @ljharb, I just made a few changes:
1. I updated shell detection to use `getent passwd` and fall back to the `$SHELL` variable -- otherwise we fall back to brute forcing it as before
2. I added a testing variable to make it always use `$SHELL` for testing, as that's easier to mock than `getent`
3. I tested against 5 different shells with Docker containers, using a minimal Ubuntu image as my base

Let me know what you think!

ljharb commented:
What is `getent`? Is it in POSIX and thus on all supported systems? My Mac doesn't have it.

ljharb commented on install.sh:
```diff
@@ -233,16 +233,54 @@ nvm_detect_profile() {
+
+  # If we're not testing, try to get shell from passwd
+  # Otherwise, try the SHELL variable
+  if [ "$NVM_TESTING" != 'yes' ]; then
```
```suggestion
  if [ "${NVM_TESTING-}" != 'yes' ]; then
```

ljharb commented on install.sh:
```diff
@@ -233,16 +233,54 @@ nvm_detect_profile() {
+  # Otherwise, try the SHELL variable
+  if [ "$NVM_TESTING" != 'yes' ]; then
+    USER_SHELL=$(getent passwd $(whoami) | cut -d: -f7)
+  elif [ -n "$SHELL" ]; then
```
```suggestion
  elif [ -n "${SHELL-}" ]; then
```

ljharb commented on install.sh:
```diff
@@ -233,16 +233,54 @@ nvm_detect_profile() {
+  # If we're not testing, try to get shell from passwd
+  # Otherwise, try the SHELL variable
+  if [ "$NVM_TESTING" != 'yes' ]; then
+    USER_SHELL=$(getent passwd $(whoami) | cut -d: -f7)
```
When I run `$(getent passwd $(whoami) | cut -d: -f7)` on my Mac, I get `-bash: getent: command not found` on stderr, which doesn't seem like something we want. What is `getent`?

steelcowboy replied:
Shoot yeah I meant to see if getent exists on Mac, apparently it doesn't :/ Back to the drawing board!

steelcowboy replied:
So looking at it, there doesn't seem to be a standard POSIX way to do this -- and the current behavior of checking various shell versions certainly isn't POSIX, as there is no $DASH_VERSION for example.

As such, I propose we do this: getent is the most reliable way on Linux machines, and it's quite standard. We could go one step further and just `grep $(whoami)` on `/etc/passwd` if we want to avoid getent.

On MacOS, I think the best we can do is hope the $SHELL variable is accurate because Apple. MacOS apparently does its own thing for storing user info, so we could either write in logic to use `getent` on Linux and whatever MacOS-equivalent on MacOS, or we can see if `getent` exists, use it if possible, fall back to `$SHELL`, and finally fall back to brute forcing. Either way, I'm confident that doing that will:
1. Make all supported shells work on Linux
2. Support at LEAST bash and zsh on MacOS, except there's still the remaining issue of MacOS not creating zshrc :) But again, that's a hurdle for a different PR

What do you think?

ljharb replied:
I'm fine with trying a bunch of different things in the install script, as long as they fail gracefully on systems where they don't work.

steelcowboy replied:
Got it! So `getent` obviously works reliably on Linux, but I'm exploring a more cross-platform solution of simply seeing what the parent process of the script is :) I think it's quite reasonable, because it's most likely that people would be installing `nvm` from their default shell.

Furthermore, if you decide to start a new `bash` session when your default shell is `zsh` simply to install `nvm`, I think it's fair to say it's the user's responsibility at that point since my impression is that opening a different type of shell just to install something is pretty non-standard. We can document it to say that running the script will "install nvm for the currently running shell" too so the behavior doesn't surprise anyone.

steelcowboy commented:
That makes sense on testing variables! The issue here we need to resolve is that we're trying to learn about system state in a testing environment, which makes sense if we can just pass a variable (e,.g. SHELL) but it's a lot harder if we want to mock, for example, `/etc/passwd`. So if we want to trust some system file over `$SHELL`, I think for the unit tests we'll have to just ignore that code path or overengineer the test case.

Do you see a better way forward?

ljharb commented:
I'd focus on what makes the production code correct, and I'm happy to help figure out how to mock it properly, rather than thinking about tests when writing the real code :-)

tg90nor commented:
Can't you just look at the `$SHELL` variable @steelcowboy ?
```
[~] ❯ bash -c 'echo $BASH_VERSION'
3.2.57(1)-release
[~] ❯ bash -c 'echo $ZSH_VERSION'

[~] ❯ echo $ZSH_VERSION
5.8
[~] ❯ bash -c 'echo $SHELL'
/usr/local/bin/zsh
```
This is on MacOS, but I expect it would work the same on Linux

vincent89291 reviewed (commented):
Explicit

ljharb commented:
Replaced by #2556.
